model/base_code_model.py

```diff
@@ -0,0 +1,158 @@
+        pass
+
+    @abstractmethod
+    def _get_progress_bar(self, log: Dict, group: str) -> Dict[str, float]:
+        pass
+
+    # ===== OPTIMIZERS =====
+
+    def configure_optimizers(self) -> Tuple[List[Optimizer], List[_LRScheduler]]:
+        if self.config.optimizer == "Momentum":
+            # using the same momentum value as in original realization by Alon
+            optimizer = SGD(
+                self.parameters(),
+                self.config.learning_rate,
+                momentum=0.95,
+                nesterov=self.config.nesterov,
+                weight_decay=self.config.weight_decay,
+            )
+        elif self.config.optimizer == "Adam":
+            optimizer = Adam(self.parameters(), self.config.learning_rate, weight_decay=self.config.weight_decay)
+        else:
+            raise ValueError(f"Unknown optimizer name: {self.config.optimizer}, try one of: Adam, Momentum")
+        scheduler = ExponentialLR(optimizer, self.config.decay_gamma)
+        return [optimizer], [scheduler]
+
+    # ===== DATALOADERS BLOCK =====
+
+    def train_dataloader(self) -> DataLoader:
+        dataloader, n_samples = create_dataloader(
+            self.config.train_data_path,
+            self.config.max_context,
+            self.config.random_context,
+            self.config.shuffle_data,
+            self.config.batch_size,
+            self.num_workers,
+        )
+        print(f"approximate number of steps for train is {ceil(n_samples / self.config.batch_size)}")
+        return dataloader
+
+    def val_dataloader(self) -> DataLoader:
+        dataloader, n_samples = create_dataloader(
+            self.config.val_data_path,
+            self.config.max_context,
+            False,
+            False,
+            self.config.test_batch_size,
+            self.num_workers,
+        )
+        print(f"approximate number of steps for val is {ceil(n_samples / self.config.test_batch_size)}")
+        return dataloader
+
+    def test_dataloader(self) -> DataLoader:
+        dataloader, n_samples = create_dataloader(
+            self.config.test_data_path,
+            self.config.max_context,
+            False,
+            False,
+            self.config.test_batch_size,
+            self.num_workers,
+        )
+        print(f"approximate number of steps for test is {ceil(n_samples / self.config.test_batch_size)}")
+        return dataloader
+
+    # ===== STEP =====
+
+    def transfer_batch_to_device(self, batch: PathContextBatch, device: torch.device) -> PathContextBatch:
+        # Dict str -> torch.Tensor [seq length; batch size * n_context]
+        for k in batch.context:
+            batch.context[k] = batch.context[k].to(self.device)
+        # [seq length; batch size]
+        batch.labels = batch.labels.to(self.device)
+        return batch
+
+    def training_step(self, batch: PathContextBatch, batch_idx: int) -> Dict:
```

Comment:
I suggest making this abstract.
1. Calculating logits may be different due to the different args (you don't need sequence length and ground truth labels for classification problem).
2. For classification problem, as I mentioned before, you don't need another method at all.
3. Different problems -- different metrics and logging to progress bar.